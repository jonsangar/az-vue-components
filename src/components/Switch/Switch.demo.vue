<template lang="pug">
  .switchDemo
    Demo(title="Uso Básico")
      .padding-x
        az-switch(v-model="checked")

    Demo(title="Desactivado")
      .padding-x
        az-switch(v-model="disabled" disabled)

    Demo(title="Tamaño")
      p.padding-x
        az-switch(v-model="size1" size="large")
      p.padding-x
        az-switch(v-model="size2")
      p.padding-x
        az-switch(v-model="size3" size="small")

    Demo(title="Color personalizado")
      .padding-x
        az-switch(v-model="color" active-color="var(--color-success)" inactive-color="var(--color-danger)")


    Demo(title="Modo celda")
      az-cell(title="title")
        az-switch(v-model="cell")

    Demo(title="Control asíncrono")
      .padding-x
        az-switch(:value="async" @input="onInput" :disabled="loading")
        az-loading.margin-left(type="bars" v-if="loading")
        p.small.color-gray4 timeout: 2 segundos
</template>

<script>
export default {
  data() {
    return {
      checked: true,
      disabled: true,
      size1: true,
      size2: true,
      size3: true,
      color: true,
      async: false,
      cell: true,
      loading: false,
    };
  },

  methods: {
    onInput(checked) {
      this.loading = true;

      setTimeout(() => {
        this.async = checked;
        this.loading = false;
      }, 2000);
    },
  },
};
</script>

<style lang="scss">
.switchDemo {
  min-height: 100vh;
  background-color: color(gray7);

  .demo__content {
    padding: 0;
  }
}
</style>
