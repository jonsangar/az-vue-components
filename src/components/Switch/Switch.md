---
device: true
---

# Switch

### Instalación

```javascript
import { Switch } from 'didor';

Vue.use(Switch);
```

---

### Uso

#### Uso básico

``` html
<az-switch v-model="checked"/>
```

#### Desactivado

``` html
<az-switch v-model="checked" disabled/>
```

#### Tamaño

``` html
<az-switch v-model="checked" size="large" disabled/>
<az-switch v-model="checked" disabled/>
<az-switch v-model="checked" size="small" disabled/>
```

#### Color personalizado

``` html
<az-switch v-model="checked" active-color="var(--color-success)" inactive-color="var(--color-danger)" />
```

#### Modo celda

``` html
<az-cell title="title">
  <az-switch v-model="checked"/>
</az-cell>
```

#### Control asíncrono

``` html
<az-switch :value="checked" @input="onInput" :disabled="loading"/>
<az-loading type="bars" v-if="loading"/>
```

``` js
export default {
  data() {
    return {
      checked: true,
      loading: false,
    };
  },

  methods: {
    async onInput(checked) {
      this.loading = true;

      await callService();

      this.loading = false;
      this.checked = checked;
    }
  }
};
```


---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| v-model | Indica el estado del switch | `any` | `false` |
| disabled | Deshabilita el switch | `boolean` | `false` |
| size | Tamaño del switch<br>`small` `large` | `String` | -- |
| active-color | Color de fondo cuando está activo | `String` | `var(--color-brand)` |
| inactiveColor | Color de fondo cuando está inactivo | `String` | `var(--color-gray7)` |
| active-value | Valor del switch cuando está activo | `any` | `false` |
| inactive-value | Valor del switch cuando está desactivado | `any` | `false` |

### Eventos

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| change | Se lanza cuando cambia el valor del switch | -         |
