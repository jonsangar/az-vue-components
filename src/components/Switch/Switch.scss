$switch-width: double() !default;
$switch-height: simple() !default;
$switch-background-color: #ffffff !default;
$switch-border: 1px solid rgba(0, 0, 0, 0.1) !default;
$switch-node-size: simple() !default;
$switch-node-background-color: #ffffff !default;
$switch-node-box-shadow: 0 3px 1px 0 rgba(0, 0, 0, 0.05), 0 2px 2px 0 rgba(0, 0, 0, 0.1),
  0 3px 3px 0 rgba(0, 0, 0, 0.05) !default;
$switch-transition-duration: 300ms !default;
$switch-on-background-color: color(brand) !default;
$switch-disabled-opacity: 0.4 !default;

.az-switch {
  $rootClass: &;

  position: relative;
  box-sizing: initial;
  display: inline-block;
  width: $switch-width;
  height: $switch-height;
  vertical-align: middle;
  cursor: pointer;
  background-color: $switch-background-color;
  border: $switch-border;
  border-radius: $switch-node-size;
  transition: background-color $switch-transition-duration cubic-bezier(0.8, 0, 0.2, 1);

  &__node {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    width: $switch-node-size;
    height: $switch-node-size;
    background-color: $switch-node-background-color;
    border-radius: 100%;
    box-shadow: $switch-node-box-shadow;
    transition: all $switch-transition-duration cubic-bezier(0.8, 0, 0.2, 1);
  }

  &__loading {
    top: 25%;
    left: 25%;
    width: 50%;
    height: 50%;
    line-height: 1;
  }

  &--on {
    background-color: $switch-on-background-color;

    #{$rootClass}__node {
      transform: translateX($switch-width - $switch-node-size);
    }
  }

  &--disabled {
    cursor: initial;
    opacity: $switch-disabled-opacity;
  }

  &--small {
    width: size(1.5);
    height: size(0.75);

    #{$rootClass}__node {
      width: size(0.75);
      height: size(0.75);
    }

    &#{$rootClass}--on #{$rootClass}__node {
      transform: translateX(size(1.5) - size(0.75));
    }
  }

  &--large {
    width: size(2.5);
    height: size(1.25);

    #{$rootClass}__node {
      width: size(1.25);
      height: size(1.25);
    }

    &#{$rootClass}--on #{$rootClass}__node {
      transform: translateX(size(2.5) - size(1.25));
    }
  }
}
