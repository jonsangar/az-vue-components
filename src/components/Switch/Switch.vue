<template lang="pug">
  .az-switch(
    :class="switchClass"
    :style="switchStyle"
    role="switch"
    :aria-checked="String(checked)"
    @click="onClick")
    .az-switch__node
</template>

<script>
export default {
  props: {
    value: {
      type: [String, Number, Boolean],
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    activeColor: {
      type: String,
      required: false,
    },
    activeValue: {
      type: [String, Boolean],
      default: true,
    },
    inactiveColor: {
      type: String,
      required: false,
    },
    inactiveValue: {
      type: [String, Boolean],
      default: false,
    },
    size: {
      type: String,
      required: false,
      validator: value => ['small', 'large'].indexOf(value) !== -1,
    },
  },

  computed: {
    checked() {
      return this.value === this.activeValue;
    },

    switchClass() {
      return [
        this.disabled ? 'az-switch--disabled' : '',
        this.checked ? 'az-switch--on' : '',
        this.size ? `az-switch--${this.size}` : '',
      ];
    },

    switchStyle() {
      return {
        backgroundColor: this.checked ? this.activeColor : this.inactiveColor,
      };
    },
  },

  methods: {
    onClick() {
      if (!this.disabled) {
        const newValue = this.checked ? this.inactiveValue : this.activeValue;
        this.$emit('input', newValue);
        this.$emit('change', newValue);
      }
    },
  },
};
</script>

<style src="./Switch.scss" lang="scss"></style>
