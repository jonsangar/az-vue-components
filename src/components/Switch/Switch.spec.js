import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Switch from './Switch.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Switch.vue', () => {
  const build = () => {
    const wrapper = shallowMount(Switch, {
      localVue,
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Switch con snapshot', () => {
    const { wrapper } = build()
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot()

    const elem = wrapper.find('.az-switch');

    expect(elem.isVisible()).toBe(true);

    expect(elem.attributes().role).toBe('switch');
  })

  it('Comprueba valores activo/inactivo por defecto', () => {
    const { wrapper } = build()
    wrapper.vm.$nextTick();

    expect(wrapper.vm.inactiveValue).toBe(false);
    expect(wrapper.vm.activeValue).toBe(true);
  })

  it('Comprueba valores personalizados de activo/incativo', () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: true,
        activeValue: 'active',
        inactiveValue: 'inactive'
      }
    })
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.vm.activeValue).toBe('active');
    expect(wrapper.vm.inactiveValue).toBe('inactive');


  })

  it('Comprueba estado "ON" cuando el valor introducido es el adecuado', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: true
      }
    }
    )
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();

    const elem = wrapper.find('.az-switch');

    expect(elem.classes()).toContain('az-switch--on');
    expect(elem.attributes()['aria-checked']).toBe('true')

  })

  it('Comprueba estado "OFF" cuando el valor introducido es el adecuado', async () => {
    const { wrapper } = build()
    wrapper.vm.$nextTick();
    // *********************
    // Probando con Boolean
    // *********************
    await wrapper.setProps({
      value: false
    });

    expect(wrapper.html()).toMatchSnapshot();

    const elem = wrapper.find('.az-switch');

    expect(wrapper.vm.checked).toBe(false);
    expect(elem.classes()).not.toContain('az-switch--on');
    expect(elem.attributes()['aria-checked']).toBe('false');
  });

  it('Comprueba que al pasar de "off" a "on" se ejecuta como es debido', async () => {
    const { wrapper } = build();
    wrapper.vm.$nextTick();

    const elem = wrapper.find('.az-switch');

    await elem.trigger('click');
    // Pasamos de inactivo a activo
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toEqual([true]);
    expect(wrapper.emitted().change).toBeTruthy();
    expect(wrapper.emitted().change[0]).toEqual([true]);
    // expect(wrapper.vm.value).toBe(true); // Se debería de emitir el y actualizar el valor [!]

  })

  it('Comprueba que al pasar de "on" a "off" se ejecuta como es debido', async () => {
    // Nota inicializamos el valor del props a la hora de montarlo (para forzar estado inicial a true)
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: true
      }
    })

    const elem = wrapper.find('.az-switch');

    await elem.trigger('click');
    // Pasamos de activa a inactivo
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toEqual([false]);
    expect(wrapper.emitted().change).toBeTruthy();
    expect(wrapper.emitted().change[0]).toEqual([false]);
    // expect(wrapper.vm.value).toBe(false); // Se debería de emitir el y actualizar el valor [!]

  })

  it('Comprueba que al pasar de "on" a "off" se ejecuta con valores personalizados', async () => {
    // Nota inicializamos el valor del props a la hora de montarlo (para forzar estado inicial a true)
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: 'active',
        activeValue: 'active',
        inactiveValue: 'inactive'
      }
    })

    const elem = wrapper.find('.az-switch');

    await elem.trigger('click');
    // Pasamos de activa a inactivo
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toEqual(['inactive']);
    expect(wrapper.emitted().change).toBeTruthy();
    expect(wrapper.emitted().change[0]).toEqual(['inactive']);
    // expect(wrapper.vm.value).toBe('inactive'); // Se debería de emitir el y actualizar el valor [!]

  })

  it('Comprueba que al pasar de "off" a "on" se ejecuta con valores personalizados', async () => {
    // Nota inicializamos el valor del props a la hora de montarlo (para forzar estado inicial a true)
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: 'inactive',
        activeValue: 'active',
        inactiveValue: 'inactive'
      }
    })

    const elem = wrapper.find('.az-switch');

    await elem.trigger('click');
    // Pasamos de activa a inactivo
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toEqual(['active']);
    expect(wrapper.emitted().change).toBeTruthy();
    expect(wrapper.emitted().change[0]).toEqual(['active']);
    // expect(wrapper.vm.value).toBe('active'); // Se debería de emitir el y actualizar el valor [!]

  })

  it('Comprueba que se aplica correctamente "activeColor"', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: true,
        activeColor: 'red'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');

    expect(elem.attributes()).toHaveProperty('style');
    expect(elem.attributes().style).toContain('background-color: red');
  });

  it('Comprueba que no se aplican estilos de "activeColor" cuando Switch está inactivo', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: false,
        activeColor: 'red'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');

    expect(elem.attributes()).not.toHaveProperty('style');
  });

  it('Comprueba que se aplica correctamente "inactiveColor"', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: false,
        inactiveColor: 'blue'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');

    expect(elem.attributes().style).toContain('background-color: blue');

  });

  it('Comprueba que no se aplican estilos de "inactiveColor" cuando Switch está activo', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        value: true,
        inactiveColor: 'blue'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');

    expect(elem.attributes()).not.toHaveProperty('style');
  });

  it('Comprueba que se aplica el estilo "size small"', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        size: 'small'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');
    expect(elem.classes()).toContain('az-switch--small');
  })

  it('Comprueba que se aplica el estilo "size large"', async () => {
    const wrapper = shallowMount(Switch, {
      localVue,
      propsData: {
        size: 'large'
      }
    });
    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const elem = wrapper.find('.az-switch');
    expect(elem.classes()).toContain('az-switch--large');
  });

  it('Comprueba que no se aplican "sizes" no válidos', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Switch.props.size.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('small')).toBe(true);
    expect(validator('large')).toBe(true);
    expect(validator('fakesize')).toBe(false);
  });


  it('Comprueba el estado "disabled"', async () => {
    const { wrapper } = build();
    wrapper.vm.$nextTick();

    await wrapper.setProps({
      disabled: true
    });

    expect(wrapper.html()).toMatchSnapshot();

    const elem = wrapper.find('.az-switch');

    expect(elem.classes()).toContain('az-switch--disabled');

    await elem.trigger('click');

    expect(wrapper.emitted().input).toBeFalsy();
    expect(wrapper.emitted().change).toBeFalsy();
  });

})
