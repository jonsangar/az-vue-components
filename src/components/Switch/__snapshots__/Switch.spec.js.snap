// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch.vue Comprueba el estado "disabled" 1`] = `
<div role="switch" aria-checked="false" class="az-switch az-switch--disabled">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba estado "OFF" cuando el valor introducido es el adecuado 1`] = `
<div role="switch" aria-checked="false" class="az-switch">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba estado "ON" cuando el valor introducido es el adecuado 1`] = `
<div role="switch" aria-checked="true" class="az-switch az-switch--on">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que no se aplican estilos de "activeColor" cuando Switch está inactivo 1`] = `
<div role="switch" aria-checked="false" class="az-switch">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que no se aplican estilos de "inactiveColor" cuando Switch está activo 1`] = `
<div role="switch" aria-checked="true" class="az-switch az-switch--on">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que se aplica correctamente "activeColor" 1`] = `
<div role="switch" aria-checked="true" class="az-switch az-switch--on" style="background-color: red;">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que se aplica correctamente "inactiveColor" 1`] = `
<div role="switch" aria-checked="false" class="az-switch" style="background-color: blue;">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que se aplica el estilo "size large" 1`] = `
<div role="switch" aria-checked="false" class="az-switch az-switch--large">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba que se aplica el estilo "size small" 1`] = `
<div role="switch" aria-checked="false" class="az-switch az-switch--small">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba renderizado del componente Switch con snapshot 1`] = `
<div role="switch" aria-checked="false" class="az-switch">
  <div class="az-switch__node"></div>
</div>
`;

exports[`Switch.vue Comprueba valores personalizados de activo/incativo 1`] = `
<div role="switch" aria-checked="false" class="az-switch">
  <div class="az-switch__node"></div>
</div>
`;
